File: nodejs/src/api_common.js
export class SpamError extends Error {
  constructor(message) {
    super(message);
    this.name = 'SpamError';
  }
}

export class RateLimitError extends Error {
  constructor(message) {
    super(message);
    this.name = 'RateLimitError';
  }
}

export class RequestFailedError extends Error {
  constructor(message) {
    super(message);
    this.name = 'RequestFailedError';
  }
}

export class ApiCommon {
  constructor(api_config) {
    this.api_config = api_config;
  }

  // common code for printing the response to a successful transaction
  print_response(response) {
    if (this.api_config.verbosity >= 1) {
      console.log(`<Response [${response.statusCode}]>`);
      if (this.api_config.verbosity >= 3) {
        console.log('x-pinterest-rid:', response.headers['x-pinterest-rid']);
        console.log(response.body);
      }
    }
  }

  // common code for printing and rethrowing the error in response to a transaction
  print_and_throw_error(error) {
    if (!error) {
      throw new RequestFailedError('unknown error');
    }
    if (!error.response) {
      console.log('error:', error);
      throw new RequestFailedError('error without a response');
    }
    if (!error.response.body) {
      console.log('error response:', error.response);
      throw new RequestFailedError('error without a response body');
    }
    if (!error.response.body.message) {
      console.log('error response body:', error.response.body);
      throw new RequestFailedError('error without a response body message');
    }
    const error_message = `request failed with reason: ${error.response.body.message}`;
    if (this.api_config.verbosity >= 1) {
      console.log(`<Response [${error.response.statusCode}]>`);
      console.log(error_message);
      if (this.api_config.verbosity >= 2) {
        console.log('x-pinterest-rid:', error.response.headers['x-pinterest-rid']);
        console.log(error.response.body);
      }
    }
    if (error.response.statusCode === 429) {
      const detail = error.response.body.message_detail || error.response.body.message;
      if (detail && detail.toLowerCase().includes('spam')) {
        throw new SpamError(detail);
      }
      throw new RateLimitError(error_message);
    }
    throw new RequestFailedError(error_message);
  }
}

File: python/src/api_common.py
class RateLimitException(Exception):
    """Raised when API emits a HTTP 429 Too Many Requests Error"""

    pass


class SpamException(Exception):
    """Raised when API emits a HTTP 429 due to a spam issue"""


class ApiCommon:
    """Common code for using the Pinterest API"""

    def __init__(self, api_config):
        self.api_config = api_config

    def _check(self, response):
        if not hasattr(response, "ok") or not hasattr(response, "reason"):
            raise TypeError("unexpected response object: " + response)

    def check(self, response):
        """Check for errors and respond appropriately."""

        self._check(response)

        # Save a human-readable status for output and error handling.
        status = (
            "ok" if response.ok else "request failed with reason: " + response.reason
        )

        # Print a short summary of the response and and error message, if necessary.
        if self.api_config.verbosity >= 1:
            print(response)
            if not response.ok:
                print(status)

        # Handle errors.
        if not response.ok:
            if self.api_config.verbosity >= 2:
                print("x-pinterest-rid:", response.headers.get("x-pinterest-rid"))
            if response.status_code == 429:
                raise RateLimitException
            raise RuntimeError(status)

        if self.api_config.verbosity >= 3:
            print("x-pinterest-rid:", response.headers.get("x-pinterest-rid"))

    def unpack(self, response):
        """
        Check for errors, retrieve the response, and respond appropriately.

        The response argument is the return value of a REST transaction
        executed by the requests module.
        """
        self._check(response)

        # Save a human-readable status for output and error handling.
        status = (
            "ok" if response.ok else "request failed with reason: " + response.reason
        )

        # Print a short summary of the response and and error message, if necessary.
        if self.api_config.verbosity >= 1:
            print(response)
            if not response.ok:
                print(status)

        # The response should always have JSON content. If the response is empty or if
        # the JSON is not valid, raise an appropriate error.
        try:
            unpacked = response.json()
        except Exception as err:
            raise RuntimeError("response does not have valid json content: " + str(err))

        # Handle errors.
        if not response.ok:
            if self.api_config.verbosity >= 2:
                print("x-pinterest-rid:", response.headers.get("x-pinterest-rid"))
                print(unpacked)
            if response.status_code == 429:
                detail = unpacked.get("message_detail") or unpacked.get("message")
                if (
                    detail and "spam" in detail.lower()
                ):  # reason for 429 response is spam
                    raise SpamException(detail)
                raise RateLimitException
            raise RuntimeError(status)

        # Continue normal processing.
        if self.api_config.verbosity >= 3:
            print("x-pinterest-rid:", response.headers.get("x-pinterest-rid"))
            print(unpacked)

        return unpacked
